fully_shard load state_dict

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90945

Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.

rohan-varma committed:
[WIP] fully_shard load state_dict

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] fully_shard load state_dict"




[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -137,6 +137,21 @@ def subtest_name(test_name_mapping, *args):
         [test_name_mapping[str(s)] if s is not None else "none" for s in args]
     )
 
+def _broadcast_state_dict(rank, state_dict):
```
not strictly needed right now but will be used in composable tests.

rohan-varma committed:
Update on "fully_shard load state_dict"




[ghstack-poisoned]

rohan-varma committed:
Update on "fully_shard load state_dict"




[ghstack-poisoned]

awgu commented on test/distributed/_composable/test_fully_shard.py:
```diff
@@ -452,44 +455,148 @@ def world_size(self) -> int:
+            self.assertEqual(p.sum(), 0)
+
+        sd = {k: v.clone() for k, v in composable_sd.items()}
+        print(f"sd keys: {list(sd.keys())}")
```
nit: Remember to remove.

awgu commented on test/distributed/_composable/test_fully_shard.py:
```diff
@@ -452,44 +455,148 @@ def world_size(self) -> int:
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_save_load_flow(self):
+        """
+        E2E test of save + load with rank0_only + CPU offload for TransformerWithSharedParams
```
In the future, will this test include different state dict types and subtest the different configs?

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -237,14 +237,20 @@ def _common_unshard_post_state_dict_hook(
-            buffers, buffer_dtypes, fsdp_state.compute_device
+    if buffers:
+        mixed_precision_enabled_for_buffers = (
+            fsdp_state._mixed_precision_enabled_for_buffers() if not _is_composable(fsdp_state)
```
To-do: We can make `_mixed_precision_enabled_for_buffers()` not be a method of `FullyShardedDataParallel` to make this not have to `if` / `else` here. We would be able to just check `fsdp_state.mxied_precision.buffer_dtype is not None` -- flexible whether that is in its own function or written like that every time.

awgu reviewed (approved):
LGTM! There a few to-dos you left for yourself. Feel free to address those before landing.

rohan-varma committed:
Update on "fully_shard load state_dict"


Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.


[ghstack-poisoned]

rohan-varma committed:
Update on "fully_shard load state_dict"


Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.


[ghstack-poisoned]

rohan-varma committed:
Update on "fully_shard load state_dict"


Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.


[ghstack-poisoned]

rohan-varma committed:
Update on "fully_shard load state_dict"


Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.


[ghstack-poisoned]

rohan-varma committed:
Update on "fully_shard load state_dict"


Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.


[ghstack-poisoned]

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -111,15 +111,18 @@ def _assert_module_states(
-            param.zero_()
-    if zero_buffers:
-        for buffer in model.buffers():
+    ctx = FSDP.summon_full_params(model) if summon_full else suppress()
```
Should we include any to-do or issue for following up on this? Or, could you remind me what the current status on this is?

rohan-varma replied:
yeah, let me file an issue summarizing it.

rohan-varma commented:
lint jobs are unrelated and failing due to clang-tidy binary hash issue.

rohan-varma commented:
@pytorchbot merge -f "CI passed"
